include/nil/blueprint/lookup_table_library.hpp

```diff
@@ -0,0 +1,142 @@
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
+// SOFTWARE.
+//---------------------------------------------------------------------------//
+
+#ifndef CRYPTO3_LOOKUP_TABLE_LIBRARY_HPP
+#define CRYPTO3_LOOKUP_TABLE_LIBRARY_HPP
+
+#include <string>
+#include <map>
+
+#include <nil/crypto3/zk/snark/arithmetization/plonk/detail/lookup_table_definition.hpp>
+#include <nil/blueprint/manifest.hpp>
+#include <nil/blueprint/assert.hpp>
+
+namespace nil {
+    namespace blueprint {
+        template <typename Type, typename = void>
+        struct component_use_lookup : std::false_type{ };
+
+        template <typename Type>
+        struct component_use_lookup<Type, 
+            typename std::enable_if<std::is_member_function_pointer<decltype(&Type::component_lookup_tables)>::value>::type> : std::true_type
+        { };
+
+        template <typename Type>
+        constexpr bool use_lookups(){
+            if(component_use_lookup<Type>::value){
+                return true;
+            }
+            return false;
+        }
+
+        template <typename Type, typename = void>
+        struct component_use_custom_lookup_tables : std::false_type{ };
+
+        template <typename Type>
+        struct component_use_custom_lookup_tables<Type, 
+            typename std::enable_if<std::is_member_function_pointer<decltype(&Type::component_custom_lookup_tables)>::value>::type> : std::true_type
+        { };
+
+        template <typename Type>
+        constexpr bool use_custom_lookup_tables(){
+            if(component_use_custom_lookup_tables<Type>::value){
+                return true;
+            }
+            return false;
+        }
+
+        template <typename BlueprintFieldType>
+        class lookup_table_library{
+            using lookup_table_definition = typename nil::crypto3::zk::snark::detail::lookup_table_definition<BlueprintFieldType>;
+            using filled_lookup_table_definition = typename nil::crypto3::zk::snark::detail::filled_lookup_table_definition<BlueprintFieldType>;
+
+            class binary_xor_table_type : public lookup_table_definition{
+            public:
+                binary_xor_table_type(): lookup_table_definition("binary_xor_table"){
+                    this->subtables["full"] = {{0,1,2}, 0, 3};
+                }
+                virtual void generate(){
+                    this->_table = {
+                        {0, 0, 1, 1},
+                        {0, 1, 0, 1},
+                        {0, 1, 1, 0}
+                    };
+                }
+            };
+        protected:
+            binary_xor_table_type binary_xor_table;
+
+            std::map<std::string, lookup_table_definition*> tables;
```

Comment:
Smart pointers here would be better, but we probably do not care much currently because this is a global object.
Currently, this has a potential to leak memory.